### PR TITLE
feat(Badge) pt.1 - math utilities for the Badge component

### PR DESCRIPTION
**Pull request by jmuzina**
## Done

Implementation and testing for two math utilities for the Badge component:
- clamp(): clamp a number between a min and max
- displayNumber(): given a number, return its string representation, restricted to a certain maximum character count. IE 1200000 -> 1.2M

In the existing badge component, this was handled with very badge-specific and difficult-to-localize functions, seen [here](https://github.com/canonical/react-components/blob/758ba66c1dbcac450fbda4dd5e060fa532fe5346/src/components/Badge/Badge.tsx#L38-L56). This PR adds the functionality needed for badge but also defines it more extensibly so it can be used by other components, like pagination, charts, etc.

## QA

- See that the tests pass.

### PR readiness check

- [x] PR should have one of the following labels:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [x] PR title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) format. 
- [x] All packages define the required scripts in `package.json`:
  - [x] All packages: `check`, `check:fix`, and `test`.
  - [x] Packages with build steps: `build` to build the package for development or distribution, `build:all` to build **all** artifacts. See [CONTRIBUTING.md](../old/CONTRIBUTING.md#24-full-artifact-builds-buildall) for details.

### CODE REVIEW

File: packages/utils/package.json [hunk 1]
```diff
@@ -36,12 +36,18 @@
     "check:biome": "biome check",
     "check:biome:fix": "biome check --write",
     "check:ts": "tsc --noEmit",
-    "test": "echo 'No tests defined yet'"
+    "test": "bun run test:vitest",
+    "test:watch": "bun run test:vitest:watch",
```
[the remark this answers is not in the document]
Comment (jmuzina) [reply packages/utils/package.json:40]: What if somebody is working specifically within one package and they want to keep an eye on whether their changes cause a test to fail?

File: packages/utils/src/index.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 export { default as casing } from "./casing.js";
 export { default as debounce } from "./debounce.js";
 export { default as invariant } from "./invariant.js";
+export * as math from "./math/index.js";
```
[the remark this answers is not in the document]
Comment (jmuzina) [reply packages/utils/src/index.ts:4]: refactored, no longer following this pattern

File: packages/utils/tsconfig.json [hunk 3]
```diff
@@ -4,5 +4,6 @@
     "baseUrl": "src",
     "types": ["node"]
   },
-  "include": ["src/**/*.ts"]
+  "include": ["src/**/*.ts"],
+  "exclude": ["src/**/*.test.ts"]
```
[the remark this answers is not in the document]
Comment (jmuzina) [reply packages/utils/tsconfig.json]: undone

File: packages/utils/src/math/index.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export { default as clamp } from "./clamp.js";
+export { default as displayNumber, displayNumberWithUnit } from "./display.js";
```
[the remark this answers is not in the document]
Comment (jmuzina) [reply packages/utils/src/math/index.ts]: done

File: packages/utils/src/math/display.tests.ts [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+    });
+  });
+
+  describe("edge cases", () => {
+    it("handles unusual input numbers", () => {
+      const testCases: TestCase[] = [
+        { input: 0, expected: "0" },
+        { input: -100, expected: "0" },
+        { input: NaN, expected: "0" },
+        { input: Infinity, expected: "99T+" },
```
[the remark this answers is not in the document]
Comment (jmuzina) [reply packages/utils/src/math/display.tests.ts]: done

File: packages/utils/vitest.config.ts
[the remark this answers is not in the document]
Comment (jmuzina) [reply packages/utils/vitest.config.ts:1]: Perhaps this can be handled as a separate issue. We already have some ["old" documentation](https://github.com/canonical/pragma/blob/1e371d836aad30fbe491a118d29c41ae5d4125c5/old/TESTING.md#setup-and-configuration) detailing this setup as a standard, which we need to replace with new standards when time allows.

**Comment by jmuzina** [comment]
Please have another look at this @advl

File: packages/utils/tsconfig.json [hunk 6]
```diff
@@ -2,7 +2,9 @@
   "extends": "@canonical/typescript-config-base",
   "compilerOptions": {
     "baseUrl": "src",
-    "types": ["node"]
+    "types": ["node", "vitest/globals"],
+    "skipLibCheck": true
```
[the remark this answers is not in the document]
Comment (jmuzina) [reply packages/utils/tsconfig.json]: removed

**Comment by jmuzina** [comment]
@advl I've updated this to reflect your suggestion to remove the clamping behavior. Please have another look.
